KAFKA-19767 - Send Share-Fetch one-node at a time for record_limit mode

ShivsundarR opened the pull request:
*What*

- After KIP-1206, when `record_limit` mode was introduced, we ideally
want no more than the #records in the `maxRecords` field in
`ShareFetchRequest`.
- Currently, the client broadcasts the share fetch requests to all nodes
which host the leaders of the partitions that it is subscribed to.
- The application thread would be woken up after the first response
arrives, but meanwhile the responses from other nodes could bring in
those many #records next and would wait in the buffer, that would mean
we are wasting the acquisition locks for these records which are
waiting.
- Instead we would want to only send the next request when we poll
again.

- PR aims to send the request to only 1 node at a time in record_limit
mode.
- We are using partition-rotation on each poll so that no partition is
starved.

There were NCSS checkstyle errors in `ShareConsumeRequestManagerTest`,
so added a few refactors there to reduce the length.

Performance
- When we have more consumers than the #partitions(i.e when real sharing
of data happens in a partition), then we are seeing the performance is
almost the same as the current approach. But when we have lesser
consumers than the #partitions, then we see a performance regression as
client is waiting for a node to return a response before it can send the
next request.
- Hence we have introduced this only for `record_limit` mode for now,
future work will be done to improve this area.

Reviewers: Andrew Schofield <aschofield@confluent.io>

ShivsundarR committed:
Implement one-node logic for record_limit mode

ShivsundarR committed:
Remove local file

ShivsundarR committed:
NCSS check fix

AndrewJSchofield commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManagerTest.java:
```diff
@@ -1834,14 +1831,7 @@ public void testWhenShareFetchResponseReturnsALeadershipChangeErrorButNoNewLeade
-                .setRecords(records)
-                .setAcquiredRecords(ShareCompletedFetchTest.acquiredRecords(1L, 1))
-                .setAcknowledgeErrorCode(Errors.NONE.code()));
+        LinkedHashMap<TopicIdPartition, ShareFetchResponseData.PartitionData> partitionData = buildPartitionDataMap(tip0, records, ShareCompletedFetchTest.acquiredRecords(1L, 1), Errors.NONE, Errors.NONE);
```
nit: This is an excessively long line. Please break it as `=`.

AndrewJSchofield commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManagerTest.java:
```diff
@@ -1853,8 +1843,7 @@ public void testWhenShareFetchResponseReturnsALeadershipChangeErrorButNoNewLeade
         Map<TopicPartition, List<ConsumerRecord<byte[], byte[]>>> partitionRecords = fetchRecords();
-        assertTrue(partitionRecords.containsKey(tp0));
-        assertFalse(partitionRecords.containsKey(tp1));
+        assertTrue(partitionRecords.containsKey(tp0) && !partitionRecords.containsKey(tp1));
```
I wonder whether this change is redundant.

AndrewJSchofield commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManagerTest.java:
```diff
@@ -1899,13 +1879,9 @@ public void testWhenShareFetchResponseReturnsALeadershipChangeErrorButNoNewLeade
-        assertEquals(1, fetchedRecords.size());
-        fetchedRecords = partitionRecords.get(tp1);
-        assertEquals(1, fetchedRecords.size());
+        assertTrue(partitionRecords.containsKey(tp0) && partitionRecords.containsKey(tp1));
```
I would leave this line unchanged as two assertions just to minimise the amount of change.

AndrewJSchofield commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManagerTest.java:
```diff
@@ -1936,16 +1912,9 @@ public void testWhenFetchResponseReturnsWithALeadershipChangeErrorAndNewLeaderIn
-                .setRecords(records)
-                .setAcquiredRecords(ShareCompletedFetchTest.acquiredRecords(1L, 1))
-                .setAcknowledgeErrorCode(Errors.NONE.code()));
+        LinkedHashMap<TopicIdPartition, ShareFetchResponseData.PartitionData> partitionData = buildPartitionDataMap(tip0, records, ShareCompletedFetchTest.acquiredRecords(1L, 1), Errors.NONE, Errors.NONE);
```
Very long again.

AndrewJSchofield commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManagerTest.java:
```diff
@@ -1958,8 +1927,7 @@ public void testWhenFetchResponseReturnsWithALeadershipChangeErrorAndNewLeaderIn
         Map<TopicPartition, List<ConsumerRecord<byte[], byte[]>>> partitionRecords = fetchRecords();
-        assertTrue(partitionRecords.containsKey(tp0));
-        assertFalse(partitionRecords.containsKey(tp1));
+        assertTrue(partitionRecords.containsKey(tp0) && !partitionRecords.containsKey(tp1));
```
And here.

AndrewJSchofield commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManagerTest.java:
```diff
@@ -1997,13 +1958,9 @@ public void testWhenFetchResponseReturnsWithALeadershipChangeErrorAndNewLeaderIn
-        assertEquals(1, fetchedRecords.size());
-        fetchedRecords = partitionRecords.get(tp1);
-        assertEquals(1, fetchedRecords.size());
+        assertTrue(partitionRecords.containsKey(tp0) && partitionRecords.containsKey(tp1));
```
And here.

AndrewJSchofield commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManagerTest.java:
```diff
@@ -2053,8 +2003,7 @@ public void testWhenLeadershipChangeBetweenShareFetchRequests(Errors error) {
         Map<TopicPartition, List<ConsumerRecord<byte[], byte[]>>> partitionRecords = fetchRecords();
-        assertTrue(partitionRecords.containsKey(tp0));
-        assertFalse(partitionRecords.containsKey(tp1));
+        assertTrue(partitionRecords.containsKey(tp0) && !partitionRecords.containsKey(tp1));
```
And here.

AndrewJSchofield commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManagerTest.java:
```diff
@@ -2137,23 +2073,9 @@ void testLeadershipChangeAfterFetchBeforeCommitAsync() {
-                        .setRecords(records)
-                        .setAcquiredRecords(ShareCompletedFetchTest.acquiredRecords(1L, 1))
-                        .setAcknowledgeErrorCode(Errors.NONE.code()));
+        LinkedHashMap<TopicIdPartition, ShareFetchResponseData.PartitionData> partitionData = buildPartitionDataMap(tip0, records, ShareCompletedFetchTest.acquiredRecords(1L, 1), Errors.NONE, Errors.NONE);
```
And here

AndrewJSchofield commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManagerTest.java:
```diff
@@ -2509,6 +2394,90 @@ void testWhenLeadershipChangedAfterDisconnected() {
+        Set<TopicPartition> partitions = new HashSet<>();
+        partitions.add(tp0);
+        partitions.add(tp1);
+        subscriptions.assignFromSubscribed(partitions);
```
I don't think this is necessarily deterministic order. I worry that this might be a new flaky test.

ShivsundarR replied:
Right, I looked into this and there are 2 ways to solve this I thought.
1. Change the test to be order-agnostic, this makes the test a bit bigger, but we can make it work.
2. Use `LinkedHashMap` in `SubscriptionState` to maintain insertion order, so that we can be deterministic in the test and in any future tests that other classes may use.

I have updated the PR with option 2 as this was just a temporary map used in `assignFromSubscibed` before updating the actual **assignment**. Does this work?

AndrewJSchofield reviewed (changes requested):
Thanks for the PR. Generally looks pretty good, but I have some formatting feedback for the test, and one comment about perhaps a test case which might sometimes fail unexpectedly.

Please could you merge latest changes from trunk into your branch. Thanks.

ShivsundarR committed:
Merge remote-tracking branch 'upstream/trunk' into KAFKA-19767-shiv

ShivsundarR committed:
Address review comments

ShivsundarR commented:
Thanks @AndrewJSchofield for the review,
I had made the refactors mainly for the checkstyle to pass, it was showing code limit exceeded(NCSS), so tried combining the asserts, added some helper functions for it to pass. 
Now I reverted the assert changes, and checkstyle still passed, so I have reverted them now.
